extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,617 @@
+          log.makeAlert(e, "Exception in KafkaSupervisor[%s]", dataSource)
+             .emit();
+        }
+
+        Thread.sleep(config.getRunPeriod());
+      }
+    }
+    catch (InterruptedException e) {
+      log.info("KafkaSupervisor[%s] interrupted, exiting.", dataSource);
+    }
+  }
+
+  private static String getRandomId()
+  {
+    final StringBuilder suffix = new StringBuilder(8);
+    for (int i = 0; i < Ints.BYTES * 2; ++i) {
+      suffix.append((char) ('a' + ((RANDOM.nextInt() >>> (i * 4)) & 0x0F)));
+    }
+    return suffix.toString();
+  }
+
+  private KafkaConsumer<byte[], byte[]> getKafkaConsumer(String bootstrapServers)
+  {
+    final Properties props = new Properties();
+    props.putAll(spec.getIoConfig().getConsumerProperties());
+
+    props.setProperty("enable.auto.commit", "false");
+    props.setProperty("metadata.max.age.ms", "10000");
+    props.setProperty("bootstrap.servers", bootstrapServers);
+    props.setProperty("group.id", String.format("kafka-supervisor-%s", supervisorRunId));
+
+    ClassLoader currCtxCl = Thread.currentThread().getContextClassLoader();
+    try {
+      Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
+      return new KafkaConsumer<>(props, new ByteArrayDeserializer(), new ByteArrayDeserializer());
+    }
+    finally {
+      Thread.currentThread().setContextClassLoader(currCtxCl);
+    }
+  }
+
+  private void updatePartitionDataFromKafka()
+  {
+    try {
+      consumer.listTopics(); // updates the consumer's list of partitions from the brokers
+    }
+    catch (Exception e) { // calls to the consumer throw NPEs when the broker doesn't respond
+      log.warn(
+          e,
+          "Unable to get partition data from Kafka for brokers [%s], are the brokers up?",
+          spec.getIoConfig().getKafkaBrokers()
+      );
+      return;
+    }
+
+    List<PartitionInfo> partitions = consumer.partitionsFor(spec.getIoConfig().getTopic());
+    int numPartitions = (partitions != null ? partitions.size() : 0);
+
+    log.debug("Found [%d] Kafka partitions for topic [%s].", numPartitions, spec.getIoConfig().getTopic());
+
+    for (int partition = 0; partition < numPartitions; partition++) {
+      int taskGroupId = getTaskGroupIdForPartition(partition);
+      if (taskGroups.get(taskGroupId) == null) {
+        log.info("Creating new task group [%d]", taskGroupId);
+        taskGroups.put(taskGroupId, new TaskGroup());
+      }
+
+      if (!taskGroups.get(taskGroupId).partitions.containsKey(partition)) {
+        log.info(
+            "New partition [%d] discovered for topic [%s], adding to task group [%d]",
+            partition,
+            spec.getIoConfig().getTopic(),
+            taskGroupId
+        );
+        taskGroups.get(taskGroupId).partitions.put(partition, new PartitionOffsets(0, 0, false));
+      }
+    }
+  }
+
+  private void updateTaskMetadata()
+  {
+    int taskCount = 0;
+    List<Task> tasks = taskStorage.getActiveTasks();
```

Comment:
hmm, this may be easier to reason about if we grab a list of active tasks at the start of each `mainLoop` iteration, and then remove things and add things as we kill and create them? maybe?